### PR TITLE
Add New Event for Customizing Chain Policy Before Certificate Validation

### LINKED ISSUE #49788: Enable Automatic Rotation of Trusted Client Certificate Chain Components
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

While I can use a combination of `ServerCertificateSelector` and `IFileProvider` today to automatically reload my server's TLS certificate (as mentioned in https://github.com/dotnet/aspnetcore/issues/32351), I cannot as easily reload the certificate(s) used to validate the client's certificate chain. I can of course write my own `ClientCertificateValidation` delegate, but I think that also means re-performing much of the logic that already exists in the [`Microsoft.AspNetCore.Authentication.Certificate`](https://www.nuget.org/packages/Microsoft.AspNetCore.Authentication.Certificate) package.

### Describe the solution you'd like

It would be very helpful to have parallels to `HttpsConnectionAdapterOptions.ServerCertificateSelector` in the `CertificateAuthenticationOptions` class, which already provides the properties `CustomTrustStore` and `AdditionalChainCertificates`. I think it would be reasonable to include selectors here too, like `CustomTrustStoreSelector` and `AdditionalChainCertificatesSelector` that provide perhaps the client certificate as input and output an `X509Certificate2Collection` object.

E.g.
```csharp
public Func<X509Certificate2, X509Certificate2Collection>? CustomTrustStoreSelector { get; set; }
```

Although the names may be a bit verbose, and perhaps there should be some connection info in there too, in case someone wants to change the CA based on the endpoint?

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wsugarman**
# Add New Event for Customizing Chain Policy Before Certificate Validation
Add new event that is invoked before a certificate is validated.

## Description
This Pull Request introduces a new event called `OnCertificateValidating` that is invoked after the `X509ChainPolicy` is built but before the `509Chain` is built (and the certificate is validated). In this way, developers can augment the policy to perhaps include certificate chain components dynamically.

Fixes #49788

### CODE REVIEW

File: src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs [hunk 1]
```diff
@@ -216,15 +216,17 @@ private X509ChainPolicy BuildChainPolicy(X509Certificate2 certificate, bool isCe
         }
         else
         {
-            if (Options.CustomTrustStore != null)
+            X509Certificate2Collection? customTrustStore = Options.CustomTrustStoreSelector?.Invoke(certificate) ?? Options.CustomTrustStore;
```
Comment (Tratcher) [inline src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs:219]:
```suggestion
            var customTrustStore = Options.CustomTrustStoreSelector?.Invoke(certificate) ?? Options.CustomTrustStore;
```

File: src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs [hunk 2]
```diff
@@ -216,15 +216,17 @@ private X509ChainPolicy BuildChainPolicy(X509Certificate2 certificate, bool isCe
...
             {
-                chainPolicy.CustomTrustStore.AddRange(Options.CustomTrustStore);
+                chainPolicy.CustomTrustStore.AddRange(customTrustStore);
             }
 
             chainPolicy.TrustMode = Options.ChainTrustValidationMode;
         }
 
-        chainPolicy.ExtraStore.AddRange(Options.AdditionalChainCertificates);
+        X509Certificate2Collection? additionalChainCertificates = Options.AdditionalChainCertificatesSelector?.Invoke(certificate) ?? Options.AdditionalChainCertificates;
```
Comment (Tratcher) [inline src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs:228]:
```suggestion
        var additionalChainCertificates = Options.AdditionalChainCertificatesSelector?.Invoke(certificate) ?? Options.AdditionalChainCertificates;
```

File: src/Security/Authentication/Certificate/src/CertificateAuthenticationOptions.cs [hunk 3]
```diff
@@ -23,18 +23,30 @@ public class CertificateAuthenticationOptions : AuthenticationSchemeOptions
     /// </summary>
     public X509Certificate2Collection CustomTrustStore { get; set; } = new X509Certificate2Collection();
 
+    /// <summary>
+    /// A callback that will be invoked to dynamically select the trusted X509 certificate components of the certificate chain.
+    /// This is higher priority than <see cref="CustomTrustStore"/>.
+    /// </summary>
+    public Func<X509Certificate2, X509Certificate2Collection>? CustomTrustStoreSelector { get; set; }
```
Comment (Tratcher) [inline src/Security/Authentication/Certificate/src/CertificateAuthenticationOptions.cs:30]: Should the HttpContext (or some other object) be passed in for context?

Comment (Tratcher) [reply src/Security/Authentication/Certificate/src/CertificateAuthenticationOptions.cs:30]: Move these to https://github.com/dotnet/aspnetcore/blob/52d1c3020ac9b0f747b3ff0c244e7ec2be6f5554/src/Security/Authentication/Certificate/src/Events/CertificateAuthenticationEvents.cs and follow the pattern there?

**Review by Tratcher — CHANGES_REQUESTED** [review]
This will need API review. See https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title= for details.

File: src/Security/Authentication/Certificate/src/CertificateAuthenticationOptions.cs (hunk 3 above)
Comment (wsugarman) [reply src/Security/Authentication/Certificate/src/CertificateAuthenticationOptions.cs]: I think this is a good idea! My latest iteration now leverages these events.

**Comment by halter73** [comment]
Thanks again for the PR and API proposal. We would love to make automatic rotation of the `CustomTrustStore` for client certificates easier in .NET 9. However, we've decided to close this PR until we reach a conclusion on #49789 and approve a more idiomatic and efficient API based on the feedback in https://github.com/dotnet/aspnetcore/issues/49788#issuecomment-1804855894. We can apply the `api-approved` label and accept a new PR. Definitely keep the branch around in case you can reuse any of it.

**Comment by wsugarman** [comment]
Makes sense to me!

**Pull request closed without merging**